enhancements/machine-api/aws-placement-groups.md

```diff
@@ -0,0 +1,543 @@
+  - To allow control of machines/workloads across partitions of a Partition placement group
+- Create Placement Groups based on user configuration when they do not exist
+- Remove Placement Groups when they are no longer required
+- Ensure Placement Groups can be removed during the cluster deprovisioning process
+- Allow appropriate usage of Partition Placement Groups across multiple MachineSets.
+
+### Non-Goals
+
+- Allowing this feature to be leveraged during installation (day 0)
+  - Users will be able to modify worker MachineSets after manifest creation if required but this will not be exposed as
+    a part of the installation configuration before manifest generation
+- Support for Placement Groups for Control Plane Machines
+  - Users may replace their control plane at a later date using future automation if they wish to have their control
+    plane within a placement group
+- Support for moving Machines between placement groups
+
+## Proposal
+
+We will introduce new fields within the `Placement` section of the AWS Machine Provider Config which will allow users
+to specify which placement group to create instances within.
+
+If the placement group does not exist, it will be created based upon the group type and, if applicable partition count provided.
+
+We will also allow end users to specify which partition their instances should be created within when appropriate.
+
+### User Stories
+
+- As an OpenShift cluster administrator, I would like to use a spread placement group to ensure that my critical
+  workloads are unlikely to suffer from hardware failures simultaneously.
+- As an operator of High-Performance Computing workloads on OpenShift, I would like to use a cluster placement group to
+  ensure that there is
+  minimal latency between instances within my cluster, allowing maximal throughput of my workloads.
+- As an operator of a large scale distributed and replicated data store, I would like to use a partition placement
+  group to ensure that different shards of my data are replicated within different fault domains within the same
+  availability zone.
+- As an OpenShift cluster administrator, I would like unneeded placement groups to be cleaned up so that I am not
+  polluting my environment with resources that are no longer in use.
+
+### API Extensions
+
+The new fields for this enhancement will be limited to the `Placement` struct within the `AWSMachineProviderConfig` in
+the Machine v1beta1 API group.
+
+The changes are outlined below:
+
+```golang
+type Placement struct {
+  // Existing Fields: Region, AvailabilityZone and Tenancy
+  ...
+
+  // GroupName specifies the name of an AWS placement group to create the Machine within.
+  // If the group specified does not exist, it will be created based on the GroupType
+  // paramenter.
+  // When a Machine is deleted, if this leaves an empty Placement Group, the group is also
+  // deleted.
+  // +optional
+  GroupName string `json:"groupName,omitempty"`
```

Comment:
If I have three machinesets and each contains an AWSMachineProviderConfig and I specify a GroupName of "BestGroupEver" in each machineset, what happens?

Reply:
That's allowed and is a valid use case. If you read through the partition placement groups section, it mentions having a partition placement group with multiple partitions and having each partition represented by the group number.

There isn't a negative effect to having multiple MachineSets with the same group name, if it already exists, it will just be used as is

Reply:
> There isn't a negative effect to having multiple MachineSets with the same group name, if it already exists, it will just be used as is

Given three different definitions of the GroupName, which of those definition is accurate?  Based on what I'm seeing here, I gather that it is racy.

Using this as both the definition PlacementGroup and the reference to which PlacementGroup to use makes the usage of this API inconsistent depending on read order.  The same resources played against two different clusters can have two different results.  this is a problem with the API.

Reply:
> Given three different definitions of the GroupName, which of those definition is accurate? Based on what I'm seeing here, I gather that it is racy.

I think you're right this would cause a problem for users as one of the MachineSets would come up successfully and the others, as we have it implemented now, would fail with errors

> As an alternative, why not have one resource that describes the PlacementGroup and simply reference those PlacementGroups from here.

This is something we could do, though it would be a lot more effort to get done (will need to review team planning/product requirements) and does add complexity for the users. A concern that has come up before is proliferation of the number of CRDs within OpenShift, does this tiny feature warrant adding a new CRD? Is that acceptable? 

I'll have a few conversations today and gather feedback, fwiw, to avoid a lot of the complexity here about preventing misconfiguration, having a new CRD is probably the best thing to do. I'm wondering if we should investigate the idea of the `AWSCluster` resource from Cluster API which gathers a lot of information like this, if that's a singleton within the cluster and has a list of placement group configs, it could manage those and the other "cluster level" infrastructure